pyclam/chaoda.py

```diff
@@ -0,0 +1,129 @@
+    scores: List[int] = [scores[i] for i in range(len(scores))]
+    return _normalize(np.asarray(scores, dtype=float), normalize)
+
+
+def parent_child(graph: Graph, normalize: str = 'gaussian') -> np.array:
+    """ Determines outlier scores for points by considering ratios of cardinalities of parent-child clusters.
+
+    The ratios are weighted by the child's depth in the tree, and are then accumulated for each point in each cluster in the graph.
+    Points with relatively low accumulated ration are the outliers.
+
+    :param graph: Graph on which to calculate outlier scores.
+    :param normalize: which normalization mode to use to get scores in a [0, 1] range.
+                      Must be one of 'linear', 'gaussian', or 'sigmoid'.
+    :return: A numpy array of outlier scores for each point in the manifold that the graph belongs to.
+    """
+    if normalize not in _NORMALIZATION_MODES:
+        raise ValueError(f'Normalization method {normalize} is undefined. Must by one of {_NORMALIZATION_MODES}.')
+
+    results: np.array = np.zeros(shape=graph.manifold.data.shape[0], dtype=float)
+    for cluster in graph:
+        ancestry = graph.manifold.ancestry(cluster)
+        for i in range(1, len(ancestry)):
+            score = float(ancestry[i-1].cardinality) / (ancestry[i].cardinality * np.sqrt(i))
+            for p in cluster.argpoints:
+                results[p] += score
+    return _normalize(results, normalize)
+
+
+def graph_neighborhood(graph: Graph, normalize: str = 'gaussian') -> np.array:
+    """ Determines outlier scores by the considering the relative graph-neighborhood of clusters.
+
+    Points in clusters with relatively small neighborhoods are the outliers.
+
+    :param  graph: Graph on which to calculate outlier scores.
+    :param normalize: which normalization mode to use to get scores in a [0, 1] range.
+               Must be one of 'linear', 'gaussian', or 'sigmoid'.
+
+    :return: A numpy array of outlier scores for each point in the manifold that the graph belongs to.
+    """
+    if normalize not in _NORMALIZATION_MODES:
+        raise ValueError(f'Normalization method {normalize} is undefined. Must by one of {_NORMALIZATION_MODES}.')
+
+    def _bft(start: Cluster) -> List[int]:
+        path_length: List[int] = list()
+        visited: Set[Cluster] = set()
+        frontier: Set[Cluster] = {start}
+        while frontier:
+            visited += frontier
+            path_length.append(len(visited))
+            frontier = {neighbor for cluster in frontier for neighbor in graph.neighbors(cluster)}
+            frontier -= visited
+        return path_length
+
+    path_lengths: Dict[Cluster, List[int]] = {cluster: _bft(cluster) for cluster in graph.clusters}
+    scores: Dict[Cluster, int] = {cluster: path_length[len(path_lengths) // 4] if len(path_lengths) > 0 else 0 for cluster, path_length in path_lengths.items()}
```

Comment:
Why divided by 4?

Reply:
a quarter of the max-path-length. It worked fine. We could learn the number based on the trends of how path-lengths increase but I did not want to tackle that complexity right now. I haven't completely thought through it